src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,681 @@
+				if (ret < 0)
+					goto err;
+			}
+		}
+	}
+
+	return 0;
+
+err:
+	sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+
+	return ret;
+}
+
+static int lib_manager_unload_module(uint32_t module_id, struct sof_man_module *mod,
+				     struct sof_man_fw_desc *desc)
+{
+	struct ext_library *ext_lib = ext_lib_get();
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	void *va_base_text = (void *)mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
+	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
+	void *va_base_rodata = (void *)mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
+	int ret_tmp, ret = 0;
+
+	st_text_size = st_text_size * CONFIG_MM_DRV_PAGE_SIZE;
+	st_rodata_size = st_rodata_size * CONFIG_MM_DRV_PAGE_SIZE;
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_text, st_text_size);
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_rodata, st_rodata_size);
+
+	/* There are modules marked as lib_code. This is code shared between several modules inside
+	 * the library. Unload all lib_code modules with last none lib_code module unload.
+	 */
+	if (!mod->type.lib_code && ext_lib->mods_exec_load_cnt > 0)
+		ext_lib->mods_exec_load_cnt--;
+
+	if (ext_lib->mods_exec_load_cnt == 0) {
+		struct sof_man_module *module_entry =
+		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+		for (size_t idx = 0; idx < desc->header.num_module_entries;
+				++idx, ++module_entry) {
+			if (module_entry->type.lib_code) {
+				ret_tmp =
+				lib_manager_unload_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+							  idx, mod, desc);
+				if (ret_tmp < 0)
+					ret = ret_tmp;
+			}
+		}
+	}
+
+	return ret;
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
+						     struct sof_man_module *mod)
+{
+	uint32_t instance_bss_size =
+		 mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count;
+	uint32_t inst_offset = instance_bss_size * CONFIG_MM_DRV_PAGE_SIZE * instance_id;
+	uint8_t *va_base =
+		(uint8_t *)(mod->segment[SOF_MAN_SEGMENT_BSS].v_base_addr + inst_offset);
+
+	tr_dbg(&lib_manager_tr,
+	       "lib_manager_get_instance_bss_address() instance_bss_size: 0x%x, pointer: 0x%x",
+	       instance_bss_size, va_base);
+	return va_base;
+}
+
+static int lib_manager_allocate_module_instance(uint32_t module_id, uint32_t instance_id,
+						uint32_t is_pages, struct sof_man_module *mod)
+{
+	int ret;
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			 * CONFIG_MM_DRV_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
```

Comment:
We also need to validate text, data, bss addresses and sizes. e.g. within bounds and cache/page aligned.